### PR TITLE
Allow learnable positional embeddings in MaskedVisionTransformer

### PR DESCRIPTION
**Pull request by guarin**
## Changes

* Allow learnable positional embeddings in MaskedVisionTransformer

## How was it tested?
* Added unit tests

### CODE REVIEW

File: tests/models/test_ModelUtils.py [hunk 1]
```diff
@@ -573,6 +574,35 @@ def test_get_named_leaf_modules() -> None:
...
+    if expected_fn is not None:
+        mock_fn.assert_called_once()
+
+
+def test_initialize_learnable_positional_embedding() -> None:
+    pos_embedding = Parameter(torch.ones(1, 1, 64) * 10)
+    utils.initialize_learnable_positional_embedding(pos_embedding)
+    assert pos_embedding.requires_grad
+    # Trunc normal should clip values to small size.
+    assert torch.all(pos_embedding.abs() < 5)
```
Comment (MalteEbner) [inline tests/models/test_ModelUtils.py:603]: Why not assert that [init.trunc_normal](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.trunc_normal_) was called?

If you want to keep it like this, I would rather assert that all are <=2, as this is the default value.

**Review by MalteEbner — COMMENTED** [review]
Looks good, but I have a few questions.

File: lightly/models/utils.py [hunk 2]
```diff
@@ -742,6 +742,55 @@ def add_stochastic_depth_to_blocks(vit: Module, prob: float = 0.0, mode="row") -
...
+
+
+def initialize_learnable_positional_embedding(pos_embedding: Parameter) -> None:
+    """Initializes a learnable positional embedding.
+
+    Args:
+        pos_embedding:
+            Positional embedding parameter.
+    """
+    init.trunc_normal_(pos_embedding, std=0.02)
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/models/utils.py:790]: It is standard for ViT: https://github.com/huggingface/pytorch-image-models/blob/cec70b6779ea81cec0ca08ee4a257b52affd235a/timm/models/vision_transformer.py#L590

File: lightly/models/utils.py [hunk 3]
```diff
@@ -742,6 +742,55 @@ def add_stochastic_depth_to_blocks(vit: Module, prob: float = 0.0, mode="row") -
...
+    """Initializes the positional embedding with the given strategy.
+
+    Args:
+        pos_embedding:
+            Positional embedding parameter.
+        strategy:
+            Positional embedding initialization strategy. Valid options are:
+            ['learn', 'sincos', 'skip']. 'learn' makes the embedding learnable,
+            'sincos' creates a fixed 2D sine-cosine positional embedding, and 'skip'
+            does not initialize the positional embedding.
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/models/utils.py:759]: I would keep strategy required as it doesn't make sense to call the function without a strategy. I would also like to keep `'skip'` to be consistent with weight initialization where there is also a skip option.

File: tests/models/test_ModelUtils.py (hunk 1 above)
Comment (guarin) [reply tests/models/test_ModelUtils.py]: I want to avoid mocking if not necessary. I changed now the test to just check that the embedding was updated, this avoids relying on implementation details of how the embedding is initialized.

File: lightly/models/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MalteEbner) [reply lightly/models/utils.py:790]: I would add a comment that this the standard init for ViT. Optional: also add the link.

File: tests/models/test_ModelUtils.py [hunk 4]
```diff
@@ -573,6 +574,37 @@ def test_get_named_leaf_modules() -> None:
...
+
+
+def test_initialize_learnable_positional_embedding() -> None:
+    pos_embedding = Parameter(torch.ones(1, 1, 64))
+    orig_pos_embedding = pos_embedding.clone()
+    utils.initialize_learnable_positional_embedding(pos_embedding)
+    # Embedding must be learnable.
+    assert pos_embedding.requires_grad
+    # Embedding must have changed.
+    assert torch.any(pos_embedding != orig_pos_embedding)
```
Comment (MalteEbner) [inline tests/models/test_ModelUtils.py:605]: This could fail randomly if the rng only gives out 0. Thus I would seed it.
Furthermore, I would rather check that e.g. at least 90% have changed (given a fixed seed).

**Review by MalteEbner — COMMENTED** [review]
Approved

File: tests/models/test_ModelUtils.py (hunk 4 above)
Comment (guarin) [reply tests/models/test_ModelUtils.py:605]: It would only fail if all values were randomly 0 which should be impossible. Let's keep the test as is.

**Review by MalteEbner — APPROVED** [review]

File: tests/models/test_ModelUtils.py (hunk 4 above)
Comment (MalteEbner) [reply tests/models/test_ModelUtils.py:605]: Ok, but if it fails because all values are randomly 0, you're gonna hear a big "I told you so!" ;)

Feel free to resolve.
